### PR TITLE
refactored, PluginCreator added

### PR DESCRIPTION
**Pull request by zerkalica**

### CODE REVIEW

**Comment by cevek** [comment]
Can you remove all .js and .d.ts from git repo

File: src/typescript.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import * as ts from 'typescript/lib/typescript'
+import {patchCreateProgram} from './patchCreateProgram'
+
+patchCreateProgram(ts)
+
+export default ts
```
Comment (cevek) [inline src/typescript.ts]: Original 'typescript/lib/typescript' returns  `module.exports = ts`, but you return `module.exports.default = ts`

Comment (cevek) [reply src/typescript.ts]: Use instead `export = ts;`

File: lib/customRequire.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+};
+Object.defineProperty(exports, "__esModule", { value: true });
+var module_1 = __importDefault(require("module"));
+function customRequireResolve(dir, module) {
+    var m = new module_1.default('');
+    m.filename = dir + '/index.js';
+    m.paths = module_1.default._nodeModulePaths(dir);
+    return module_1.default._resolveFilename(module, m);
+}
+exports.customRequireResolve = customRequireResolve;
```
Comment (cevek) [inline lib/customRequire.js]: Replace to `resolve.sync`

File: src/typescript.ts (hunk 1 above)
Comment (cevek) [reply src/typescript.ts]: Or do it more functional `export = patchCreateProgram(ts)`

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import * as ts from 'typescript/lib/typescript'
```
Comment (cevek) [inline src/index.ts]: This index file is not needed

File: src/PluginCreator.ts [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        const module: PluginFactory | {default: PluginFactory} = require(modulePath)
+
+        return typeof (module as any).default === 'function'
+            ? (module as any).default
+            : module
+    }
+
+    createTransformers = (main: TransformerHost) => this._createTransformers(main)
+
+    private _createTransformers(main: TransformerHost): ts.CustomTransformers {
```
Comment (cevek) [inline src/PluginCreator.ts]: Why do you need a private method with binded public?

File: src/watchGuard.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+export = require('typescript/lib/watchGuard.js')
```
Comment (cevek) [inline src/watchGuard.ts]: Can you fix it to es6 import

File: src/PluginCreator.ts [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    (program: ts.Program, options?: PluginConfig): TransformerPlugin
+} | {
+    type: 'opts'
+    (opts: ts.CompilerOptions, options?: PluginConfig): TransformerPlugin
+} | {
+    type: 'checker'
+    (opts: ts.TypeChecker, options?: PluginConfig): TransformerPlugin
+}
+
+export type TransformerHost = ts.LanguageService | ts.Program | ts.CompilerOptions | ts.TypeChecker
```
Comment (cevek) [inline src/PluginCreator.ts]: You transfer to `creator.createTransformers(program)` only `program`
Why do you need `TransformerHost`?

File: src/PluginCreator.ts (hunk 4 above)
Comment (zerkalica) [reply src/PluginCreator.ts]: To reduce function size, created in constructor. May be it is overoptimization.

File: src/patchCreateProgram.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
+import * as ts from 'typescript'
+import {PluginCreator} from './PluginCreator'
+
+export interface BaseHost {
```
Comment (cevek) [inline src/patchCreateProgram.ts]: Why do you need this interface?

File: src/patchCreateProgram.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        rootNames: ReadonlyArray<string>,
+        options: ts.CompilerOptions,
+        host?: ts.CompilerHost,
+        oldProgram?: ts.Program
+    ): ts.Program
+    versionMajorMinor: string
+}
+
+export function patchCreateProgram<Host extends BaseHost>(
+    tsm: Host
```
Comment (cevek) [inline src/patchCreateProgram.ts]: Why is not just `tsm: typeof ts`

File: src/index.ts (hunk 3 above)
Comment (zerkalica) [reply src/index.ts]: Originally main was a "lib/typescript.js" in package.json. Now "main": "lib/index.js". Better to change it back in package.json or keep standard index.js?

File: src/PluginCreator.ts [hunk 9]
```diff
@@ -149,9 +149,7 @@ export class PluginCreator {
             : module
     }
 
-    createTransformers = (main: TransformerHost) => this._createTransformers(main)
-
-    private _createTransformers(main: TransformerHost): ts.CustomTransformers {
+    createTransformers: (TransformerHost) => ts.CustomTransformers = (main: TransformerHost) => {
```
Comment (cevek) [inline src/PluginCreator.ts]: But why so hard?

File: src/patchCreateProgram.ts (hunk 8 above)
Comment (zerkalica) [reply src/patchCreateProgram.ts]: Conflicts with tsserverlibrary and typescriptServices declarations

File: src/PluginCreator.ts (hunk 6 above)
Comment (zerkalica) [reply src/PluginCreator.ts]: PluginCreator can be used as separate helper in [rollup-plugin-typescript2](https://github.com/ezolenko/rollup-plugin-typescript2/blob/master/src/ioptions.ts#L26) and other hosts.

File: src/patchCreateProgram.ts (hunk 8 above)
Comment (cevek) [reply src/patchCreateProgram.ts]: Ok, so can fix it to 
```
export type BaseHost = Pick<typeof ts, 'createProgram' | 'versionMajorMinor'>
```
to be more confident that if ts team will change `createProgram` interface we will get an error

File: src/watchGuard.ts (hunk 5 above)
Comment (zerkalica) [reply src/watchGuard.ts]: Are imports ok? May be resolve modules dynamically via resolve module, like in src/tsc.ts ?

File: src/PluginCreator.ts [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+    transform: string
+    type?: FactoryType
+    after?: boolean
+    before?: boolean
+    afterDeclaration?: boolean
+    [options: string]: any;
+} | (ts.PluginImport & {
+    [options: string]: any
+})
+
```
Comment (cevek) [inline src/PluginCreator.ts]: Can you add `name?: string` to main interface and remove `[options: string]: any;` cause your interface just transforms into just `any`
After that you will see couple of bugs below :)

Comment (zerkalica) [reply src/PluginCreator.ts]: Can't use pattern matching here. 
This not works: ``` {name: void, transform: string} | {name: string} ```
Changed to single type.

File: src/PluginCreator.ts [hunk 11]
```diff
@@ -171,15 +171,8 @@ export class PluginCreator {
...
-                if (config.afterDeclaration) chain.afterDeclaration.push(plugin)
-                if (
-                    config.before
-                    || (
-                        config.before === undefined
-                        && config.after === undefined
-                        && config.afterDeclaration === undefined
-                    )
-                ) chain.before.push(plugin)
+                else if (config.afterDeclaration) chain.afterDeclaration.push(plugin)
```
Comment (zerkalica) [inline src/PluginCreator.ts]: Plugin can provide multiple transformers for after, before, etc.

File: src/PluginCreator.ts [hunk 12]
```diff
@@ -148,7 +148,7 @@ export class PluginCreator {
             : module
     }
 
-    createTransformers: (TransformerHost) => ts.CustomTransformers = (main: TransformerHost) => {
+    createTransformers(main: TransformerHost) {
```
Comment (zerkalica) [inline src/PluginCreator.ts]: In some cases i need to deletegate createTransformers method, for example rollup-plugin-typescript2.

File: src/PluginCreator.ts [hunk 13]
```diff
@@ -1,209 +1,208 @@
-import * as ts from 'typescript'
-import * as resolve from 'resolve'
-import * as path from 'path'
-import compareVersions from 'compare-versions'
+import * as ts from 'typescript';
+import * as resolve from 'resolve';
+import * as path from 'path';
+import compareVersions from 'compare-versions';
+import { parseConfigFileTextToJson } from 'typescript/lib/tsserverlibrary';
```
Comment (zerkalica) [inline src/PluginCreator.ts]: Unused import

File: src/PluginCreator.ts [hunk 14]
```diff
@@ -1,209 +1,208 @@
...
+    after?: TsTransformerFactory;
+    afterDeclaration?: TsTransformerFactory;
+};
+
+export type FactoryRet = TransformerPlugin | TsTransformerFactory;
+export type LSPattern = (ls: ts.LanguageService, config?: PluginConfig) => FactoryRet;
+export type ProgramPattern = (program: ts.Program, config?: PluginConfig) => FactoryRet;
+export type CompilerOptionsPattern = (opts: ts.CompilerOptions, config?: PluginConfig) => FactoryRet;
+export type TypeCheckerPattern = (opts: ts.TypeChecker, config?: PluginConfig) => FactoryRet;
+export type DefaultPattern = (context: ts.TransformationContext, program?: ts.Program) => ts.Transformer<ts.SourceFile>;
```
Comment (zerkalica) [inline src/PluginCreator.ts]: Why not ProgramPattern is default?
How much transforms use ``` context: ts.TransformationContext, program?: ts.Program ``` signature?

File: src/PluginCreator.ts (hunk 13 above)
Comment (zerkalica) [reply src/PluginCreator.ts]: Provide prettier config for your codestyle.

File: src/PluginCreator.ts [hunk 15]
```diff
@@ -1,209 +1,208 @@
...
-                return factory(this.program, options)
-        }
+export function preparePluginsFromCompilerOptions(plugins: any): PluginConfig[] {
+    if (!plugins) return [];
+    // old transformers system
+    if (plugins.length === 1 && plugins[0].customTransformers) {
+        const { before = [], after = [] } = plugins[0].customTransformers;
+        return [
+            ...before.map(item => ({ tranform: item, before: true })),
+            ...after.map(item => ({ tranform: item, after: true })),
```
Comment (zerkalica) [inline src/PluginCreator.ts]: Don't forget afterDeclaration?

File: src/PluginCreator.ts [hunk 16]
```diff
@@ -1,209 +1,208 @@
...
+        const pluginObj: PluginConfig = {
+            type: 'ls',
+            name: '',
+            transform: '',
+            after: true,
+            before: true,
+            afterDeclaration: true,
+        };
+        const possibleKeys = Object.keys(pluginObj);
+        for (let i = 0; i < configs.length; i++) {
```
Comment (zerkalica) [inline src/PluginCreator.ts]: May be for of?

File: src/patchCreateProgram.ts [hunk 17]
```diff
@@ -1,5 +1,5 @@
 import * as ts from 'typescript'
-import {PluginCreator} from './PluginCreator'
+import {PluginCreator, preparePluginsFromCompilerOptions} from './PluginCreator'
```
Comment (zerkalica) [inline src/patchCreateProgram.ts]: May be move preparePluginsFromCompilerOptions to patchCreateProgram.ts or incapsulate it into PluginCreator.createTransformers?

File: src/PluginCreator.ts (hunk 14 above)
Comment (zerkalica) [reply src/PluginCreator.ts]: But really this interface not used in transformers without some wrappers:

https://github.com/firede/ts-transform-graphql-tag/blob/master/src/transformer.ts
```ts
export default function(): ts.TransformerFactory<ts.SourceFile> {
  return (context: ts.TransformationContext): ts.Transformer<ts.SourceFile> => {
    return (sourceFile: ts.SourceFile) => ts.visitNode(sourceFile, getVisitor(context))
  }
}
```

https://github.com/kimamula/ts-transformer-keys/blob/master/transformer.ts
```ts
export default function transformer(program: ts.Program): ts.TransformerFactory<ts.SourceFile> {
  return (context: ts.TransformationContext) => (file: ts.SourceFile) => visitNodeAndChildren(file, program, context);
}
```

https://github.com/longlho/ts-transform-css-modules/blob/master/src/transform.ts#L106
```ts
export default function(opts: Opts) {
    require('css-modules-require-hook')(opts)
    return (ctx: ts.TransformationContext): ts.Transformer<ts.SourceFile> => {
        return (sf: ts.SourceFile) => ts.visitNode(sf, visitor(ctx, sf))
    }
}
```

File: src/PluginCreator.ts [hunk 18]
```diff
@@ -0,0 +1,208 @@
...
+export type DefaultPattern = (context: ts.TransformationContext, program?: ts.Program) => ts.Transformer<ts.SourceFile>;
+export type PluginFactory = LSPattern | ProgramPattern | CompilerOptionsPattern | TypeCheckerPattern | DefaultPattern;
+
+function patchEmitFiles(host: any): ts.TransformerFactory<ts.SourceFile>[] {
+    const oldEmitFiles = host.emitFiles;
+    if (oldEmitFiles.__patched instanceof Array) {
+        oldEmitFiles.__patched.length = 0;
+        return oldEmitFiles.__patched;
+    }
+
```
Comment (cevek) [inline src/PluginCreator.ts]: Can you refactor this code to be more clear, I really don't understand what happens here
Why there is a `oldEmitFiles.__patched`?

Comment (zerkalica) [reply src/PluginCreator.ts]: createTransformers can be called many times with same ts instance.
Without __patched, on second createTransformers call -  new transformers will be ignored and used from oldCreateTransformers.

File: src/PluginCreator.ts (hunk 14 above)
Comment (cevek) [reply src/PluginCreator.ts]: Ok, you're right, but then we need also to add new pattern for transformers without factory wrapper:
 `raw`
`opts` -> plugin config 
`compilerOptions` -> compileroptions 

https://github.com/dsherret/ts-nameof/blob/8dec4b18c10c95460ef4c55b154a83aa792bcb06/src/transformation/transformerFactory.ts#L7
```ts
const transformerFactory: ts.TransformerFactory<ts.SourceFile> = context => {
    return file => visitSourceFile(file, context) as ts.SourceFile;
};
```

Comment (zerkalica) [reply src/PluginCreator.ts]: Yes, need separate adapter for transformers without factory. But my question was about default common pattern. Is ts-nameof example so common pattern to be default?

Comment (cevek) [reply src/PluginCreator.ts]: Fix default to program pattern

File: src/index.ts [hunk 19]
```diff
@@ -1,3 +0,0 @@
-import ts from './typescript'
```
Comment (zerkalica) [inline src/index.ts]: Why? Not all bundlers awaits typescript as module name.

In rollup-plugin-typescript2:

```ts
import patchedTypescript from 'ttypescript'
return {
  plugins: [
    tsPlugin({typescript: patchedTypescript})
  ]
```

Comment (cevek) [reply src/index.ts]: I'll fix main to `lib/typescript.js`

File: src/patchCreateProgram.ts [hunk 20]
```diff
@@ -15,9 +15,8 @@ export function patchCreateProgram<Host extends BaseHost>(
         oldProgram?: ts.Program
     ): ts.Program {
         const program = originCreateProgram(rootNames, options, host, oldProgram)
-
-        const creator = new PluginCreator(
-            (program.getCompilerOptions() as any).plugins,
+        const pluginCreator = new PluginCreator(
```
[the remark this answers is not in the document]
Comment (zerkalica) [reply src/patchCreateProgram.ts]: Not a problem, if program is a new instance on each originCreateProgram call.

[the remark this answers is not in the document]
Comment (cevek) [reply src/patchCreateProgram.ts]: I don't follow you. What the problem if program always is new instance?

[the remark this answers is not in the document]
Comment (zerkalica) [reply src/patchCreateProgram.ts]: Problem if same instance.  But here is no problem, my mistake.

```ts
program = {
  emit
}
```

**Comment by zerkalica** [comment]
Can you check last typescript example test? Patched createProgram not called

**Comment by cevek** [comment]
You need to replace test engine from jest to something less hacky for example ava
jest replaces native `require` to own implementation
The problem is patched `ts` is not equal to `ts` inside of `transpileModule` - it is non patched original version. 
I don't know how this might be happen, so don't want waste time to digg into this shit

**Comment by cevek** [comment]
Example works if you call `node ../lib/tsc test.ts` from `example` dir

File: src/patchCreateProgram.ts [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+    return tsm.parseJsonConfigFileContent(result.config, tsm.sys, projectDir, undefined, configFileNamePath);
+}
+
+function preparePluginsFromCompilerOptions(plugins: any): PluginConfig[] {
+    if (!plugins) return [];
+    // old transformers system
+    if (plugins.length === 1 && plugins[0].customTransformers) {
+        const { before = [], after = [] } = plugins[0].customTransformers;
+        return [
+            ...before.map((item: string) => ({ tranform: item, before: true })),
```
Comment (paranoidjk) [inline src/patchCreateProgram.ts]: This typo breaks our project, please merge https://github.com/cevek/ttypescript/pull/12

Comment (cevek) [reply src/patchCreateProgram.ts]: @paranoidjk Fixed and published ttypescript@1.3.0

**Pull request merged**